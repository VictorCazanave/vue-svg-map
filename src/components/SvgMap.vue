<script setup lang="ts">
import type { SVGAttributes } from 'vue'
import type { Map, Location } from '../types'

const { locationAttributes = () => ({}) }  = defineProps<{
  map: Map
  locationAttributes?: SVGAttributes | ((location: Location, index: number) => SVGAttributes)
}>()
</script>

<template>
	<svg
		xmlns="http://www.w3.org/2000/svg"
		:viewBox="map.viewBox"
		:aria-label="map.label"
    class="svg-map"
  >
		<slot name="before" />
		<path
			v-for="(location, index) in map.locations"
			:key="location.id"
			:id="location.id"
			:d="location.path"
			:aria-label="location.name"
			class="svg-map__location"
			v-bind="{
        ...$attrs,
        ...(typeof locationAttributes === 'function' ? locationAttributes(location, index) : locationAttributes)
      }"
		/>
		<slot name="after" />
	</svg>
</template>
