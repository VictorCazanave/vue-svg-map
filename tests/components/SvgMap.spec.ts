import { mount } from '@vue/test-utils'
import type { VueWrapper } from '@vue/test-utils'
import type { Map, Location } from '../../src/types'
import map from './mocks/map'
import SvgMap from '../../src/components/SvgMap.vue'

describe('SvgMap component', () => {
  let wrapper: VueWrapper

  beforeEach(() => {
		wrapper = mount(SvgMap, { props: { map } })
  })

	it('renders svg map with default attributes of locations', () => {
		expect(wrapper.element).toMatchSnapshot()
	})

	it('renders svg map with static custom attributes of locations', async () => {
    const locationAttributes = { tabindex: 'tabindex', role: 'role' }
		await wrapper.setProps({ locationAttributes })

		expect(wrapper.element).toMatchSnapshot()
	})

	it('renders svg map with dynamic custom attributes of locations', async () => {
    const getLocationAttributes = (location: Location, index: number) => ({ 'name': location.name, 'index': index })
		await wrapper.setProps({ locationAttributes: getLocationAttributes })

		expect(wrapper.element).toMatchSnapshot()
	})

	it('renders custom svg map', async () => {
		const customMap: Map = {
			...map,
			locations: map.locations.map(location => ({
				...location,
				id: `custom-${location.id}`,
				path: `custom path of ${location.name}`,
			})),
		}
		await wrapper.setProps({ map: customMap })

		expect(wrapper.element).toMatchSnapshot()
	})
})
