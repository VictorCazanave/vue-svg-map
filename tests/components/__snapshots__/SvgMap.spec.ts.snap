// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SvgMap component > renders custom svg map 1`] = `
<svg
  aria-label="label"
  class="svg-map"
  viewBox="viewBox"
  xmlns="http://www.w3.org/2000/svg"
>
  
  
  
  <path
    aria-label="name0"
    class="svg-map__location"
    d="custom path of name0"
    id="custom-id0"
  />
  <path
    aria-label="name1"
    class="svg-map__location"
    d="custom path of name1"
    id="custom-id1"
  />
  <path
    aria-label="name2"
    class="svg-map__location"
    d="custom path of name2"
    id="custom-id2"
  />
  
  
  
</svg>
`;

exports[`SvgMap component > renders svg map with default attributes of locations 1`] = `
<svg
  aria-label="label"
  class="svg-map"
  viewBox="viewBox"
  xmlns="http://www.w3.org/2000/svg"
>
  
  
  
  <path
    aria-label="name0"
    class="svg-map__location"
    d="path0"
    id="id0"
  />
  <path
    aria-label="name1"
    class="svg-map__location"
    d="path1"
    id="id1"
  />
  <path
    aria-label="name2"
    class="svg-map__location"
    d="path2"
    id="id2"
  />
  
  
  
</svg>
`;

exports[`SvgMap component > renders svg map with dynamic custom attributes of locations 1`] = `
<svg
  aria-label="label"
  class="svg-map"
  viewBox="viewBox"
  xmlns="http://www.w3.org/2000/svg"
>
  
  
  
  <path
    aria-label="name0"
    class="svg-map__location"
    d="path0"
    id="id0"
    index="0"
    name="name0"
  />
  <path
    aria-label="name1"
    class="svg-map__location"
    d="path1"
    id="id1"
    index="1"
    name="name1"
  />
  <path
    aria-label="name2"
    class="svg-map__location"
    d="path2"
    id="id2"
    index="2"
    name="name2"
  />
  
  
  
</svg>
`;

exports[`SvgMap component > renders svg map with static custom attributes of locations 1`] = `
<svg
  aria-label="label"
  class="svg-map"
  viewBox="viewBox"
  xmlns="http://www.w3.org/2000/svg"
>
  
  
  
  <path
    aria-label="name0"
    class="svg-map__location"
    d="path0"
    id="id0"
    role="role"
    tabindex="tabindex"
  />
  <path
    aria-label="name1"
    class="svg-map__location"
    d="path1"
    id="id1"
    role="role"
    tabindex="tabindex"
  />
  <path
    aria-label="name2"
    class="svg-map__location"
    d="path2"
    id="id2"
    role="role"
    tabindex="tabindex"
  />
  
  
  
</svg>
`;
